### DISCUSSION TITLE
Using ProgressRing

### DISCUSSION THREAD

**Discussion by jeromelaban** [unoplatform/uno#10211] [Q&A]
From @nikolayvpavlov in https://github.com/unoplatform/uno/issues/8264#issuecomment-1286950326

> @jeromelaban , I **did** follow the directions in the link. If you try to use that, you get "Cannot build with both Uno.WinUI and Uno.UI nuget packages referenced." when compiling the project.
> 
> Community Toolkit can be used only with a pre-released version of another package: [Uno.CommunityToolkit.WinUI.UI.Controls](https://www.nuget.org/packages/Uno.CommunityToolkit.WinUI.UI.Controls.DataGrid/7.1.100-dev.15.g12261e2626?_src=template).
> 
> The documentation is outdated, and incorrect.

Can you detail to the exact steps that you followed? I do not believe it is incorrect but rather it's likely not explaining how to select packages precisely enough.

**Comment by nikolayvpavlov** [comment]
@jeromelaban , тthanks for picking this up.  I want to use DataGrid of Windows Community Toolkit in a Uno application that runs on WinUI / Windows App Sdk and WASM.

For this purpose, I followed the documentation and added package Uno.Microsoft.Toolkit.Uwp.UI.Controls.DataGrid (https://www.nuget.org/packages/Uno.Microsoft.Toolkit.Uwp.UI.Controls.DataGrid) to the project.  Then, trying to compile the project, I got error "Cannot build with both Uno.WinUI and Uno.UI nuget packages referenced." when trying to build the project.

Therefore, I used package [Uno.CommunityToolkit.WinUI.UI.Controls.DataGrid](https://www.nuget.org/packages/Uno.CommunityToolkit.WinUI.UI.Controls.DataGrid) mentioned at the top of the [documentation](https://platform.uno/docs/articles/uno-community-toolkit.html?tabs=tabid-winui).  It is available as a pre-release only.  With it, I am able to successfully compile the project.  However, if I use it, then the ProgressRing does not work.  The project compiles, but I get an exclamation mark icon on its place only at run-time.

Please let me know if I can help with more details.

**Reply by jeromelaban** [reply]
Thanks. The WCT documentation explains how to select the proper packages variants, as detailed at the bottom of the first tab. Is it because it should be detailed before?

For the ProgressRing, [you can follow this](https://platform.uno/docs/articles/features/progressring.html?q=progress%20ring#using-the-microsoftuixamlcontrolsprogressring). The app logs also indicate what you need to do to get it working.

**Reply by nikolayvpavlov** [reply]
@jeromelaban , documentation is not correct, because it refers to both the release version of the packages (which do not work) and the pre-release ones, which do work.

However, if you include the prerelease [Uno.CommunityToolkit.WinUI.UI.Controls.DataGrid](https://www.nuget.org/packages/Uno.CommunityToolkit.WinUI.UI.Controls.DataGrid) package, the ProgressRing does not work.  If you try then to add Uno.Lottie package, as the link you give me advises, you again get this error: "Cannot build with both Uno.WinUI and Uno.UI nuget packages referenced."

**Reply by jeromelaban** [reply]
> @jeromelaban , documentation is not correct, because it refers to both the release version of the packages (which do not work) and the pre-release ones, which do work.

I see, so the issue is only because the WinUI packages are in prerelease. We have not yet published a stable version of those, as the upstream source we used at the time was also in prerelease, which we will sometime in the future. You can use those packages safely.

> However, if you include the prerelease [Uno.CommunityToolkit.WinUI.UI.Controls.DataGrid](https://www.nuget.org/packages/Uno.CommunityToolkit.WinUI.UI.Controls.DataGrid) package, the ProgressRing does not work. If you try then to add Uno.Lottie package, as the link you give me advises, you again get this error: "Cannot build with both Uno.WinUI and Uno.UI nuget packages referenced."

For lottie, it's the same process as the community toolkit, you need to use the [Uno.WinUI.Lottie package](https://platform.uno/docs/articles/features/Lottie.html?tabs=winui).

**Comment by nikolayvpavlov** [comment]
Now this works.  However, the documentation is incorrect.  It states:

> On WASM, Xamarin.Android, Xamarin.iOS and Xamarin.macOS, you'll need the following packages:
> o Uno.UI.Lottie
> 

While in the WASM project, obviously we must use **Uno.WinUI.Lottie**.

**Reply by jeromelaban** [reply]
This statement is in the UWP tab, and you're using a WinUI app (It references `Uno.WinUI.*` packages)

**Reply by nikolayvpavlov** [reply]
Well, this is very unclear.  Since each app is in a separate project and uses different packages, the documentation must clearly state what should be used when.  

I know WinUI is still in infancy and has many problems itself, but Uno should have its dependencies documented better.

**Reply by jeromelaban** [reply]
Sure. What would make it clearer?

**Reply by nikolayvpavlov** [reply]
Perhaps in the [documentation of Lottie ](https://platform.uno/docs/articles/features/Lottie.html) should have a note like this:

> If any of your target platforms is WinUI, you must reference package Uno.WinUI.Lottie in all other platforms: WASM, Xamarin.Android, Xamarin.iOS and Xamarin.macOS.  If you do not target WinUI, but only UWP, you must reference package Uno.UI.Lottie.

I am sure you will suggest a better writing than the above. :-)

But, what happens if one wants to support both WinUI and UWP, or isn't that not possible now?

**Reply by jeromelaban** [reply]
Thanks. Supporting both UWP and WinUI is possible, but it's necessarily easy to do.

This is mainly because, it's not possible to cross-target XAML, but also because of the namespaces changes. Using C# 10's global using makes it a bit easier, but mostly it's best to target WinUI.
